adserver/decisionengine/backends.py

```diff
@@ -393,7 +400,26 @@ def select_flight(self):
             # Choose a flight based on the impressions needed
             flight_range = []
             total_clicks_needed = 0
+
             for flight in possible_flights:
+
+                # Apply niche targeting only when any flight has it.
+                # This is to track whether we should do expensive distance queries.
+                if (
+                    flight.niche_targeting
+                    and "ethicalads_ext.embedding" in settings.INSTALLED_APPS
+                ):
+                    if not self.niche_weights:
+                        from ethicalads_ext.embedding.utils import (
+                            get_niche_weights,
+                        )  # noqa
+
+                        self.niche_weights = get_niche_weights(
```

Comment:
Unless I'm reading this wrong, won't this get the distance for every advertiser? It'd be really nice if the embedding was just cached on the flight (or directly linked with a FK so `select_related` just works). Then this check would just be a computation of getting the distance between the URL's embedding and the flight's embedding. Is that correct?

Reply:
Yea, we have a few options here -- but in general a flight doesn't have a single embedding, since it can have multiple ads that each have their own. We can in theory "average" them in some fashion, but that isn't what we're doing currently.

In the future we can definitely change how this works a bit more, but for now I'm mostly just trying to reduce overall queries when we aren't doing targeting, so we can start testing it in prod and see how slow it is.

Reply:
Alright. I think we'll have to calculate a flight embedding eventually but we can at least get a start with this.